Adding Branch and Bound algorithm

runn1ng opened the pull request:
https://github.com/bitcoinjs/coinselect/issues/10

In the end, I used the C++ code from core PR as a reference ( https://github.com/bitcoin/bitcoin/pull/10637 ), not the scala code.

It seems to be working, and performing rather well! Although it matches only in <10% of cases in the simulation, the reduction of the fees is significant.

The only question is what to take as a feedback. In the original paper and in the scala code that goes with it, random shuffle + accumulative is being used. In this module simulation, min+accumulative has the best results. (Even with the total costs from https://github.com/bitcoinjs/coinselect/pull/11 )

*However*, when I tried to implement the "minimal" fallback into the scala code, the utxo set in the big example exploded, which caused larger total cost, *plus* the BnB search was significantly slower.

As I wrote in the linked pull request, I will try to port the big example into this code, what will happen.

runn1ng commented:
Code comments are welcome :) it's a big chunk of code. The comments in there are almost directly from the bitcoin-core PR

I will also try to port the bitcoin-core tests in the PR here.

runn1ng commented:
(both the scala algorithm and the C++ implementations are MIT licence, as is this code :) )

Xekyo commented on branchandbound.js:
```diff
@@ -0,0 +1,143 @@
+
+    return utils.finalize(inputs, outputs, feeRate)
+  } else {
+    const fee = feeRate * utxos.reduce(function (a, x) {
```
Just in case this isn't happening elsewhere, be sure that you don't forget accounting for the fees of the transaction overhead of 10 bytes and requested outputs to the target of the selection!

runn1ng replied:
Hm. Where is that hapenning in the achow101 PR code? I don't see it. (But maybe it happens before it goes into wallet.cpp `CWallet::SelectCoinsMinConf`)

runn1ng replied:
Nope, the C++ doesn's seem to do that either.

It is not done in the coinselection.cpp - https://github.com/achow101/bitcoin/blob/19761499d5657b1f272a2fb04185dacf218fb104/src/wallet/coinselection.cpp

And it's not done in wallet.cpp - here the nvalue is just sum of the outputs
https://github.com/achow101/bitcoin/blob/19761499d5657b1f272a2fb04185dacf218fb104/src/wallet/wallet.cpp#L2437

nvalueselect is just nvalue here (nFeeRet is 0 in the first pass, and BnB is used only in the first pass)
https://github.com/achow101/bitcoin/blob/19761499d5657b1f272a2fb04185dacf218fb104/src/wallet/wallet.cpp#L2503

Here the SelectCoins is called, which calls (through some layers) BnB with the same target
https://github.com/achow101/bitcoin/blob/19761499d5657b1f272a2fb04185dacf218fb104/src/wallet/wallet.cpp#L2555

runn1ng replied:
I think it's actually a bug in the c++ code, I have added a comment to the PR, we'll see

runn1ng replied:
It's a bug in the c++ code, now corrected. Anyway though, I noticed something weird.

When I corrected the error and add the overhead, the algorithm performs slightly *worse*. When I run, on the same randomly generated data, both the BnB with incorrect target and BnB with correct target, the BnB with incorrect targets always has better results. The difference is small, but consistent.

I have no idea why.

I will anyway add some tests finally, maybe there is some other subtle bug that I haven't found.

runn1ng replied:
Oooh the issue is in the simulation, I think it allows negative fees. Which is what happens.

Xekyo replied:
It could be that now that the transaction size is properly estimated, the allowance window for the exact match actually is too big. You might want to try with a "costOfChange" that is actually only the saved change output's cost. That would be the lower bound on the saved cost, but may cause the algorithm to find fewer matches or require more inputs to find matches.

runn1ng replied:
I can replicate the results with your code too -
 https://github.com/Xekyo/CoinSelectionSimulator/issues/5

runn1ng replied:
[for other readers: the discussion continued into the bitcoind pull request]

Xekyo commented on branchandbound.js:
```diff
@@ -0,0 +1,143 @@
+var utils = require('./utils')
+
+var maxTries = 1000000
```
My gut feeling was 100k, you have 1M here. That might be very long on a huge wallet in an unfortunate UTXO pool and target combination. Might want to check what amount of time that allows.

runn1ng replied:
A typo I think! Thanks

runn1ng replied:
nope, you have 1M here too :))

https://github.com/Xekyo/CoinSelectionSimulator/blob/master/src/main/scala/one/murch/bitcoin/coinselection/StackEfficientTailRecursiveBnB.scala#L9

runn1ng replied:
but yeah 100k should be enough

runn1ng replied:
I have left 1M here; in my experience it's very fast anyway, even in javascript on browser/in node

runn1ng commented:
I haven't done the tests yet; as I said "I will also try to port the bitcoin-core tests in the PR here", which I will do today :)

dcousens commented:
Sorry for the delay here, reading now.

runn1ng commented:
Np, lots of comments here :))

On Jul 6, 2017 8:46 AM, "Daniel Cousens" <notifications@github.com> wrote:

> Sorry for the delay here, reading now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoinjs/coinselect/pull/13#issuecomment-313310687>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGZ8XRZA5BUk1_X3jtX53zJzV9zyXTvks5sLIKsgaJpZM4OLeIj>
> .
>

dcousens commented on branchandbound.js:
```diff
@@ -0,0 +1,143 @@
+  })
+
+  var selected = search(effectiveUtxos, outAccum, costOfChange)
+  if (selected != null) {
```
`!==`

runn1ng replied:
`standard-js` is already in there.

`!= null` is fine for it, since it checks both undefined and null. Maybe it would be better to somehow return "empty" value from the "inner" function more explicitly (to avoid "null !== undefined" shenanigans), than just `return;`?

runn1ng replied:
I originally thought about throwing Errors, but that's also kind of ugly

runn1ng replied:
I will leave this like this

dcousens commented:
Looks good to me,  at least for merge/ongoing research/testing :) - maybe not release until everything is tested.

dcousens commented on branchandbound.js:
```diff
@@ -0,0 +1,143 @@
+  var costPerInput = utils.inputBytes({}) * feeRate
+  var costOfChange = costPerInput + costPerOutput
+
+  var outAccum = utils.sumOrNaN(outputs)
```
`outAccum` could be `NaN` here,  is that safe?
`search` doesn't appear to handle it gracefully

runn1ng replied:
I was not sure how would the NaNs come to the outputs and how to handle them exactly, frankly, so I just delegated it to "do it later" :)

dcousens replied:
Handling them gracefully will be required for release/publish :+1: ,  but no issue for now

dcousens replied:
`NaN`s appear (intentionally,  returned by `sumOrNaN`) if any expected field is `undefined`, or `null`, or not an integer.
It proliferates `NaN` throughout the math quickly such that an error can be caught,  and is unusable as a Bitcoin value.

runn1ng commented:
It's already in there.

!= null is fine for standard.js, since it checks both undefined and null

On Jul 6, 2017 8:55 AM, "Daniel Cousens" <notifications@github.com> wrote:

> *@dcousens* commented on this pull request.
> ------------------------------
>
> In branchandbound.js
> <https://github.com/bitcoinjs/coinselect/pull/13#discussion_r125823509>:
>
> > +  if (!isFinite(utils.uintOrNaN(feeRate))) return {}
> +
> +  var costPerOutput = utils.outputBytes({}) * feeRate
> +  var costPerInput = utils.inputBytes({}) * feeRate
> +  var costOfChange = costPerInput + costPerOutput
> +
> +  var outAccum = utils.sumOrNaN(outputs)
> +
> +  var effectiveUtxos = calculateEffectiveValues(utxos, feeRate).filter(function (x) {
> +    return x.effectiveValue > 0
> +  }).sort(function (a, b) {
> +    return b.effectiveValue - a.effectiveValue
> +  })
> +
> +  var selected = search(effectiveUtxos, outAccum, costOfChange)
> +  if (selected != null) {
>
> !==
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoinjs/coinselect/pull/13#pullrequestreview-48248053>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGZ8YvzYETuL00qqlhLMXjRGLpUsrl5ks5sLITUgaJpZM4OLeIj>
> .
>

runn1ng commented:
Hey, I was not really able to look at the code (holiday :)), will go through the notes etc

runn1ng commented:
I have now added factor for change of cost

runn1ng commented on branchandbound.js:
```diff
@@ -0,0 +1,143 @@
+module.exports = function branchAndBound (utxos, outputs, feeRate) {
+  if (!isFinite(utils.uintOrNaN(feeRate))) return {}
+
+  var costPerOutput = utils.outputBytes({}) * feeRate
```
TODO: as written, this does not take segwit/different input/ouptut sizes into account

runn1ng commented on branchandbound.js:
```diff
@@ -0,0 +1,143 @@
+
+  var costPerOutput = utils.outputBytes({}) * feeRate
+  var costPerInput = utils.inputBytes({}) * feeRate
+  var costOfChange = Math.floor((costPerInput + costPerOutput) * factor)
```
this doesn't take segwit into account; but neither does `utils.finalize`

runn1ng commented:
Hm, I thought that for tests, I will copy blackjack tests, since they should be similar... but there are not blackjack tests :sob: :cry: ok then

runn1ng commented:
Hm, I now have no idea, what to return in BnB algorithm as fee, when the algorithm fails.

I will probably return just 0, since the strategy doesn't return anything meaningful for this.

runn1ng commented:
I have added some tests (mostly copying from accumulative.json) and fixed NaN issues.

It should be working.

However, the costOfChange/costPerInput/costPerOutput will need to have a revision with segwit.

runn1ng commented:
Hm. I don't know how to display the missed branches in the coverage...

dcousens commented:
@runn1ng if the fee has no meaningful result,  maybe `NaN`?

runn1ng commented:
OK, bcash is finally solved (what a time sink), so I have time to return back to work on this :)

karel-3d commented:
I see bitcoin core have added a waste metric and bunch of other stuff to BnB

mahnunchik commented:
Any news?

karelbilek commented:
lol. Not from me :) I stopped working on any bitcoin-related stuff, 0 mood to return

mahnunchik commented:
@karelbilek what happened?
